concept/type/impl/TypeImpl.java

```diff
@@ -185,6 +191,11 @@ public final boolean isAbstract() {
                     .getTypeIsAbstractRes().getAbstract();
         }
 
+        @Override
+        public boolean isRelationType() {
+            return false;
+        }
```

Comment:
We don't need this one. The root one is the correct one, and sufficient.

Reply:
This one **is** the root one, right?

Reply:
@haikalpribadi  is referring to the fact that we have duplication between `TypeImpl::isRelationType` and `TypeImpl.Remote::isRelationType`.

We can remove the duplication by moving the implementation up to the `Type` interface as default methods. What do you think of that @alexjpwalker ?

Reply:
`Type.Remote` does not extend `Type`. Therefore this one is the root one.

I am of the opinion that interface defaults are a hack in Java and should be generally avoided unless there is a good reason. Yes, we end up with some code duplication between `Type` and `Type.Remote` but that seems more like a limitation of Java when applied to our type hierarchy than anything else.

The reason is that interfaces are designed to be as dumb as possible. They should give a high-level overview of what the implementing classes are capable of, not actually tell them how to implement themselves.

Reply:
> I am of the opinion that interface defaults are a hack in Java and should be generally avoided unless there is a good reason

Yes, I agree, that they're not ideal. I try to avoid them too. But in this case, we have a good reason to use them: to avoid duplicating code multiple times, @alexjpwalker.

So as @lolski said:

> We can remove the duplication by moving the implementation up to the Type interface as default methods.

Reply:
Fixed in 3a1f1aa182cf3506c89d3d2e43e21fa689834aec